File: app/modules/aes-gcm/android/src/main/java/io/rootlens/aesgcm/AesGcmModule.kt
package io.rootlens.aesgcm

import android.util.Base64
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.io.File
import java.io.FileOutputStream
import java.nio.ByteBuffer
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

/**
 * Native AES-256-GCM (Expo Module 形式)。root-lens v0.1.0 task 18 の RN bridge 版を
 * Expo Module に書き直したもの。仕様書 §6.1 (TP 登録 E2EE channel) で使用。
 *
 * javax.crypto (Android builtin) を使う。追加依存ゼロ。
 * ARMv8 の AES hardware 命令を自動利用。
 */
class AesGcmModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("AesGcmBridge")

    /**
     * バイナリペイロード構築 + AES-256-GCM 暗号化を native で一括実行。
     * 5MB級コンテンツが JS↔Native bridge を通過しない。
     *
     * plaintext: [4B meta_len BE][metadata][content_file_bytes]
     * output:    [suite_id(1B)][encap_key_len(2B BE)][encap_key][nonce(12B)][ciphertext+tag]
     */
    AsyncFunction("buildAndEncryptPayload") {
      contentFilePath: String,
      metadataJson: String,
      requestKeyBase64: String,
      encapKeyBase64: String,
      aadString: String,
      outputFilePath: String,
      promise: Promise ->

      try {
        val key = Base64.decode(requestKeyBase64, Base64.NO_WRAP)
        val encapKey = Base64.decode(encapKeyBase64, Base64.NO_WRAP)
        val aad = aadString.toByteArray(Charsets.UTF_8)
        val metaBytes = metadataJson.toByteArray(Charsets.UTF_8)
        val content = File(contentFilePath).readBytes()

        val metaLen = ByteBuffer.allocate(4).putInt(metaBytes.size).array()
        val plaintext = metaLen + metaBytes + content

        val nonce = ByteArray(12).also { SecureRandom().nextBytes(it) }
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        cipher.updateAAD(aad)
        val ciphertext = cipher.doFinal(plaintext)

        val suiteId: Byte = 0x01  // X25519-AES-256-GCM
        val encapKeyLen = ByteBuffer.allocate(2).putShort(encapKey.size.toShort()).array()

        FileOutputStream(outputFilePath).use { out ->
          out.write(byteArrayOf(suiteId))
          out.write(encapKeyLen)
          out.write(encapKey)
          out.write(nonce)
          out.write(ciphertext)
        }

        val totalSize = 1 + 2 + encapKey.size + nonce.size + ciphertext.size
        promise.resolve(mapOf("size" to totalSize))
      } catch (e: Throwable) {
        promise.reject("BUILD_ENCRYPT_ERROR", e.message ?: "buildAndEncryptPayload failed", e)
      }
    }

    /** AES-256-GCM file → file 暗号化 (AAD 付き)。Bridge を大容量データが通らない。 */
    AsyncFunction("encryptFile") {
      inputPath: String,
      outputPath: String,
      keyBase64: String,
      aadBase64: String,
      promise: Promise ->

      try {
        val key = Base64.decode(keyBase64, Base64.NO_WRAP)
        val aad = Base64.decode(aadBase64, Base64.NO_WRAP)
        val plaintext = File(inputPath).readBytes()

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"))
        cipher.updateAAD(aad)
        val ciphertextWithTag = cipher.doFinal(plaintext)

        File(outputPath).outputStream().use { it.write(ciphertextWithTag) }

        promise.resolve(mapOf(
          "nonce" to Base64.encodeToString(cipher.iv, Base64.NO_WRAP),
          "size" to ciphertextWithTag.size,
        ))
      } catch (e: Throwable) {
        promise.reject("AES_ENCRYPT_FILE_ERROR", e.message ?: "encryptFile failed", e)
      }
    }

    /** AES-256-GCM file → file 復号 (AAD 付き)。 */
    AsyncFunction("decryptFile") {
      inputPath: String,
      outputPath: String,
      keyBase64: String,
      nonceBase64: String,
      aadBase64: String,
      promise: Promise ->

      try {
        val key = Base64.decode(keyBase64, Base64.NO_WRAP)
        val nonce = Base64.decode(nonceBase64, Base64.NO_WRAP)
        val aad = Base64.decode(aadBase64, Base64.NO_WRAP)
        val ciphertext = File(inputPath).readBytes()

        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        cipher.updateAAD(aad)
        val plaintext = cipher.doFinal(ciphertext)

        File(outputPath).writeBytes(plaintext)
        promise.resolve(outputPath)
      } catch (e: Throwable) {
        promise.reject("AES_DECRYPT_FILE_ERROR", e.message ?: "decryptFile failed", e)
      }
    }
  }
}

File: app/modules/aes-gcm/ios/AesGcmModule.swift
import ExpoModulesCore
import CryptoKit
import Foundation

// AES-256-GCM Expo Module — root-lens RN bridge 版を Expo Module に書き直したもの。
// 仕様書 §6.1 (TP 登録 E2EE) で使用。Android Kotlin 版と入出力 schema 一致。

public class AesGcmModule: Module {
  public func definition() -> ModuleDefinition {
    Name("AesGcmBridge")

    AsyncFunction("buildAndEncryptPayload") { (
      contentFilePath: String,
      metadataJson: String,
      requestKeyBase64: String,
      encapKeyBase64: String,
      aadString: String,
      outputFilePath: String,
      promise: Promise
    ) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          guard let key = Data(base64Encoded: requestKeyBase64) else {
            throw AesGcmError.invalidBase64("requestKey")
          }
          guard let encapKey = Data(base64Encoded: encapKeyBase64) else {
            throw AesGcmError.invalidBase64("encapKey")
          }
          guard let aad = aadString.data(using: .utf8) else {
            throw AesGcmError.invalidEncoding("aad")
          }
          guard let metaBytes = metadataJson.data(using: .utf8) else {
            throw AesGcmError.invalidEncoding("metadata")
          }
          let content = try Data(contentsOf: URL(fileURLWithPath: contentFilePath))

          // plaintext: [4B meta_len BE][metadata][content]
          var metaLen = UInt32(metaBytes.count).bigEndian
          var plaintext = Data(bytes: &metaLen, count: 4)
          plaintext.append(metaBytes)
          plaintext.append(content)

          // AES-256-GCM with AAD
          let symmetricKey = SymmetricKey(data: key)
          let nonce = AES.GCM.Nonce()
          let sealedBox = try AES.GCM.seal(
            plaintext,
            using: symmetricKey,
            nonce: nonce,
            authenticating: aad
          )

          // wire format v1: [suite_id(1B)][encap_key_len(2B BE)][encap_key][nonce(12B)][ct+tag]
          let suiteId: UInt8 = 0x01
          var encapKeyLen = UInt16(encapKey.count).bigEndian

          var output = Data()
          output.append(suiteId)
          output.append(Data(bytes: &encapKeyLen, count: 2))
          output.append(encapKey)
          output.append(contentsOf: sealedBox.nonce)
          output.append(sealedBox.ciphertext)
          output.append(sealedBox.tag)

          try output.write(to: URL(fileURLWithPath: outputFilePath))
          promise.resolve(["size": output.count])
        } catch {
          promise.reject("BUILD_ENCRYPT_ERROR", error.localizedDescription)
        }
      }
    }

    AsyncFunction("encryptFile") { (
      inputPath: String,
      outputPath: String,
      keyBase64: String,
      aadBase64: String,
      promise: Promise
    ) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          guard let key = Data(base64Encoded: keyBase64) else {
            throw AesGcmError.invalidBase64("key")
          }
          guard let aad = Data(base64Encoded: aadBase64) else {
            throw AesGcmError.invalidBase64("aad")
          }
          let plaintext = try Data(contentsOf: URL(fileURLWithPath: inputPath))

          let symmetricKey = SymmetricKey(data: key)
          let sealedBox = try AES.GCM.seal(plaintext, using: symmetricKey, authenticating: aad)

          // Android cipher.doFinal() = ciphertext + tag (16 byte) で揃える
          var ciphertextWithTag = Data(sealedBox.ciphertext)
          ciphertextWithTag.append(sealedBox.tag)
          try ciphertextWithTag.write(to: URL(fileURLWithPath: outputPath))

          let nonceBase64 = Data(sealedBox.nonce).base64EncodedString()
          promise.resolve(["nonce": nonceBase64, "size": ciphertextWithTag.count])
        } catch {
          promise.reject("AES_ENCRYPT_FILE_ERROR", error.localizedDescription)
        }
      }
    }

    AsyncFunction("decryptFile") { (
      inputPath: String,
      outputPath: String,
      keyBase64: String,
      nonceBase64: String,
      aadBase64: String,
      promise: Promise
    ) in
      DispatchQueue.global(qos: .userInitiated).async {
        do {
          guard let key = Data(base64Encoded: keyBase64) else {
            throw AesGcmError.invalidBase64("key")
          }
          guard let nonceData = Data(base64Encoded: nonceBase64) else {
            throw AesGcmError.invalidBase64("nonce")
          }
          guard let aad = Data(base64Encoded: aadBase64) else {
            throw AesGcmError.invalidBase64("aad")
          }
          let ciphertextWithTag = try Data(contentsOf: URL(fileURLWithPath: inputPath))

          let symmetricKey = SymmetricKey(data: key)
          let nonce = try AES.GCM.Nonce(data: nonceData)

          let tagSize = 16
          guard ciphertextWithTag.count >= tagSize else {
            throw AesGcmError.dataTooShort
          }
          let ciphertext = ciphertextWithTag.prefix(ciphertextWithTag.count - tagSize)
          let tag = ciphertextWithTag.suffix(tagSize)

          let sealedBox = try AES.GCM.SealedBox(nonce: nonce, ciphertext: ciphertext, tag: tag)
          let plaintext = try AES.GCM.open(sealedBox, using: symmetricKey, authenticating: aad)

          try plaintext.write(to: URL(fileURLWithPath: outputPath))
          promise.resolve(outputPath)
        } catch {
          promise.reject("AES_DECRYPT_FILE_ERROR", error.localizedDescription)
        }
      }
    }
  }
}

private enum AesGcmError: LocalizedError {
  case invalidBase64(String)
  case invalidEncoding(String)
  case dataTooShort

  var errorDescription: String? {
    switch self {
    case .invalidBase64(let field): return "Invalid base64 for \(field)"
    case .invalidEncoding(let field): return "Invalid UTF-8 encoding for \(field)"
    case .dataTooShort: return "Ciphertext too short (must include 16-byte GCM tag)"
    }
  }
}
